src/lib.rs

```diff
@@ -509,98 +562,122 @@ impl FailPoint {
-    /// It does not otherwise change any existing fail point configuration.
-    ///
-    /// The format of `FAILPOINTS` is `failpoint=actions;...`, where
-    /// `failpoint` is the name of the fail point. For more information
-    /// about fail point actions see the [`cfg`](fn.cfg.html) function and
-    /// the [`fail_point`](macro.fail_point.html) macro.
-    ///
-    /// `FAILPOINTS` may configure fail points that are not actually defined. In
-    /// this case the configuration has no effect.
-    ///
-    /// This function should generally be called prior to running a test with fail
-    /// points, and afterward paired with [`teardown`](#method.teardown).
-    ///
-    /// # Panics
-    ///
-    /// Panics if an action is not formatted correctly.
-    pub fn setup() -> Self {
-        // Cleanup first, in case of previous failed/panic'ed test scenarios.
-        let scenario_guard = SCENARIO.lock().unwrap_or_else(|e| e.into_inner());
-        let mut registry = scenario_guard.registry.write().unwrap();
-        Self::cleanup(&mut registry);
-
-        let failpoints = match env::var("FAILPOINTS") {
-            Ok(s) => s,
-            Err(VarError::NotPresent) => return Self { scenario_guard },
-            Err(e) => panic!("invalid failpoints: {:?}", e),
-        };
-        for mut cfg in failpoints.trim().split(';') {
-            cfg = cfg.trim();
-            if cfg.is_empty() {
-                continue;
-            }
-            let (name, order) = partition(cfg, '=');
-            match order {
-                None => panic!("invalid failpoint: {:?}", cfg),
-                Some(order) => {
-                    if let Err(e) = set(&mut registry, name.to_owned(), order) {
-                        panic!("unable to configure failpoint \"{}\": {}", name, e);
-                    }
-                }
-            }
-        }
-        Self { scenario_guard }
+    /// Deregister the current thread to this failpoints registry.
+    pub fn deregister_current(&self) {
+        let id = thread::current().id();
+        REGISTRY_GROUP.write().unwrap().remove(&id);
     }
 
-    /// Tear down the fail point system.
-    ///
-    /// Clears the configuration of all fail points. Any paused fail
-    /// points will be notified before they are deactivated.
-    ///
-    /// This function should generally be called after running a test with fail points.
-    /// Calling `teardown` without previously calling `setup` results in a no-op.
-    pub fn teardown(self) {
-        drop(self)
+    /// Clean up registered fail points in this registry.
+    pub fn cleanup(&self) {
+        let mut registry = self.registry.write().unwrap();
+        cleanup(&mut registry);
     }
+}
+
+lazy_static::lazy_static! {
+    static ref REGISTRY_GROUP: RwLock<HashMap<thread::ThreadId, Arc<RwLock<Registry>>>> = Default::default();
+    static ref REGISTRY_GLOBAL: FailPointRegistry = Default::default();
```

Comment:
When using in parallel tests, each test may concurrently call `setup`, `teardown`, `cfg` and `remove` to modify the `REGISTRY_GLOBAL` and mix up the fail point configuration for each test, especially when testing multi-thread module like `raftstore` in `tikv`, where we can't inject `current_fail_group` and `register_current` in these threads and only can use the `REGISTRY_GLOBAL` and every test will modify it concurrently.

Reply:
global and local registry do not affect each other, @NingLin-P and I agreed on this issue